sql/core/src/main/scala/org/apache/spark/sql/execution/basicOperators.scala

```diff
@@ -224,6 +225,56 @@ case class Limit(limit: Int, child: SparkPlan)
 
 /**
  * :: DeveloperApi ::
+ * Take the first limit elements. and the limit can be any number less than Integer.MAX_VALUE.
+ * If it is terminal and is invoked using executeCollect, it probably cause OOM if the
+ * records number is large enough. Not like the Limit clause, this operator will not change
+ * any partitions of its child operator.
+ */
+@DeveloperApi
+case class LargeLimit(limit: Int, child: SparkPlan)
+  extends UnaryNode {
+  /** We must copy rows when sort based shuffle is on */
+  private def sortBasedShuffleOn = SparkEnv.get.shuffleManager.isInstanceOf[SortShuffleManager]
+
+  override def output: Seq[Attribute] = child.output
+
+  override def executeCollect(): Array[Row] = child.executeTake(limit)
+
+  protected override def doExecute(): RDD[InternalRow] = {
+    val rdd = if (sortBasedShuffleOn) {
+      child.execute().map(_.copy()).persist(StorageLevel.MEMORY_AND_DISK)
+    } else {
+      child.execute().persist(StorageLevel.MEMORY_AND_DISK)
+    }
+
+    // We assume the maximize record number in a partition is less than Integer.MAX_VALUE
+    val partitionRecordCounts = rdd.mapPartitions({ iterator =>
+      Iterator(iterator.count(_ => true))
+    }, true).collect()
+
+    var totalSize = 0
+    // how many records we have to take from each partition
+    val requiredRecordCounts = partitionRecordCounts.map { count =>
```

Comment:
Will it be more efficient to use loop?  For example: you have 1000 partition count (100, 4, 5, 700, 10 ...), and limit number is 10.  If to use loop, you will do the calculation once. But if to choose map, it will do 1000 times. 
Besides, maybe we can save the storage space for requiredRecordCounts.